File: backend/tier1/trajectory_checker.py
"""Tier 1: Trajectory Checker

Detects:
- Implausible weight change velocity (too much gain/loss per day)
- Height decrease (height should never decrease)
- Physiological bounds violation
"""
from typing import List, Dict, Tuple
from datetime import datetime
from config import (
    WEIGHT_MIN_KG, WEIGHT_MAX_KG, HEIGHT_MIN_CM, HEIGHT_MAX_CM,
    ZSCORE_EXTREME
)
from tier1.zscore import calculate_zscore, get_median_weight


def check_physiological_bounds(
    value: float, obs_type: str, age_months: float, gender: str
) -> List[Dict]:
    """Check if value is within physiological possibility."""
    alerts = []

    if obs_type == "weight":
        if value < WEIGHT_MIN_KG:
            alerts.append({
                "alert_type": "PHYSIO_BOUNDS",
                "severity": "critical",
                "message": f"Weight {value} kg is below minimum physiological threshold ({WEIGHT_MIN_KG} kg).",
            })
        elif value > WEIGHT_MAX_KG:
            alerts.append({
                "alert_type": "PHYSIO_BOUNDS",
                "severity": "critical",
                "message": f"Weight {value} kg exceeds maximum pediatric threshold ({WEIGHT_MAX_KG} kg).",
            })

        # Check extreme z-score
        z = calculate_zscore(value, age_months, gender, "weight")
        if z is not None and abs(z) > ZSCORE_EXTREME:
            alerts.append({
                "alert_type": "AGE_WEIGHT_MISMATCH",
                "severity": "high",
                "message": (
                    f"Weight {value} kg has z-score {z:+.1f} for age {age_months:.0f}m ({gender}). "
                    f"This is beyond ±{ZSCORE_EXTREME} SD, which is extremely unusual."
                ),
            })

    elif obs_type == "height":
        if value < HEIGHT_MIN_CM:
            alerts.append({
                "alert_type": "PHYSIO_BOUNDS",
                "severity": "critical",
                "message": f"Height {value} cm is below minimum threshold ({HEIGHT_MIN_CM} cm).",
            })
        elif value > HEIGHT_MAX_CM:
            alerts.append({
                "alert_type": "PHYSIO_BOUNDS",
                "severity": "critical",
                "message": f"Height {value} cm exceeds maximum pediatric threshold ({HEIGHT_MAX_CM} cm).",
            })

        z = calculate_zscore(value, age_months, gender, "height")
        if z is not None and abs(z) > ZSCORE_EXTREME:
            alerts.append({
                "alert_type": "AGE_HEIGHT_MISMATCH",
                "severity": "high",
                "message": (
                    f"Height {value} cm has z-score {z:+.1f} for age {age_months:.0f}m ({gender}). "
                    f"This is beyond ±{ZSCORE_EXTREME} SD."
                ),
            })

    return alerts


def check_weight_trajectory(
    observations: List[Tuple[int, float, datetime]],
    age_months: float,
    gender: str,
) -> List[Dict]:
    """
    Check weight change velocity between consecutive measurements.

    Args:
        observations: List of (obs_id, weight_kg, effective_date) sorted by date
        age_months: Current age in months
        gender: 'male' or 'female'
    """
    alerts = []
    if len(observations) < 2:
        return alerts

    sorted_obs = sorted(observations, key=lambda x: x[2])

    # Age-adjusted max daily weight change (simplified)
    # Neonates can gain ~30g/day, older children ~10-20g/day
    median_w = get_median_weight(age_months, gender)
    # Max ~3% of body weight per day as extreme threshold
    max_daily_change = max(0.15, median_w * 0.03)

    for i in range(1, len(sorted_obs)):
        prev_id, prev_w, prev_date = sorted_obs[i - 1]
        curr_id, curr_w, curr_date = sorted_obs[i]

        days = max((curr_date - prev_date).total_seconds() / 86400, 0.01)
        change = curr_w - prev_w
        daily_change = abs(change) / days

        if daily_change > max_daily_change and days >= 0.5:
            direction = "gained" if change > 0 else "lost"
            alerts.append({
                "alert_type": "IMPLAUSIBLE_VEL",
                "severity": "high",
                "message": (
                    f"Patient {direction} {abs(change):.2f} kg over {days:.1f} days "
                    f"({daily_change:.2f} kg/day). Maximum expected: {max_daily_change:.2f} kg/day. "
                    f"Verify measurement accuracy or check for fluid changes."
                ),
                "observation_ids": [prev_id, curr_id],
                "change_kg": round(change, 3),
                "days": round(days, 1),
            })

    return alerts


def check_height_decrease(
    observations: List[Tuple[int, float, datetime]]
) -> List[Dict]:
    """Check if height decreases between consecutive measurements (should never happen)."""
    alerts = []
    if len(observations) < 2:
        return alerts

    sorted_obs = sorted(observations, key=lambda x: x[2])

    for i in range(1, len(sorted_obs)):
        prev_id, prev_h, prev_date = sorted_obs[i - 1]
        curr_id, curr_h, curr_date = sorted_obs[i]

        if curr_h < prev_h - 0.5:  # Allow 0.5cm measurement tolerance
            alerts.append({
                "alert_type": "HEIGHT_DECREASE",
                "severity": "warning",
                "message": (
                    f"Height decreased from {prev_h} cm to {curr_h} cm "
                    f"({prev_h - curr_h:.1f} cm loss). Height should not decrease in children. "
                    f"Check for measurement error or different measurement position."
                ),
                "observation_ids": [prev_id, curr_id],
            })

    return alerts

File: backend/tier1/trajectory-checker.js
/**
 * Tier 1: Trajectory Checker
 * R006: Implausible weight velocity (>5% per day)
 * R007: Weight exceeded physiological bounds
 * R008: Age-weight mismatch (z-score beyond ±5)
 * R009: Height decrease (impossible for growing child)
 * R010: BMI extreme outlier
 */
const config = require('../config');
const { weightForAgeZScore, heightForAgeZScore, ageInMonths } = require('./zscore');

function checkTrajectory(observation, patient, history) {
  const alerts = [];

  // R007: Physiological bounds
  if (observation.type === 'weight') {
    if (observation.value < config.WEIGHT_MIN_KG || observation.value > config.WEIGHT_MAX_KG) {
      alerts.push({
        alert_type: 'R007_PHYSIOLOGICAL_BOUNDS',
        severity: config.SEVERITY.CRITICAL,
        message: `Weight ${observation.value} kg outside physiological bounds (${config.WEIGHT_MIN_KG}-${config.WEIGHT_MAX_KG} kg)`,
        explanation: `A weight of ${observation.value} kg is outside the range of physiologically possible values for any pediatric patient.`
      });
    }
  }

  if (observation.type === 'height') {
    if (observation.value < config.HEIGHT_MIN_CM || observation.value > config.HEIGHT_MAX_CM) {
      alerts.push({
        alert_type: 'R007_PHYSIOLOGICAL_BOUNDS',
        severity: config.SEVERITY.CRITICAL,
        message: `Height ${observation.value} cm outside physiological bounds (${config.HEIGHT_MIN_CM}-${config.HEIGHT_MAX_CM} cm)`,
        explanation: `A height of ${observation.value} cm is outside the range of physiologically possible values for any pediatric patient.`
      });
    }
  }

  // R008: Z-score extremes
  if (observation.type === 'weight' && patient) {
    const zResult = weightForAgeZScore(observation.value, patient.date_of_birth, patient.gender, new Date(observation.effective_date));
    if (zResult.zscore < config.ZSCORE_EXTREME_LOW || zResult.zscore > config.ZSCORE_EXTREME_HIGH) {
      alerts.push({
        alert_type: 'R008_ZSCORE_EXTREME',
        severity: config.SEVERITY.CRITICAL,
        message: `Weight z-score = ${zResult.zscore} (${zResult.interpretation}). Extreme outlier for ${ageInMonths(patient.date_of_birth, new Date(observation.effective_date))}-month-old ${patient.gender}`,
        explanation: `A z-score of ${zResult.zscore} means this weight is extremely far from the WHO reference median. This is very likely a data entry error unless there is documented clinical justification.`
      });
    }
  }

  if (observation.type === 'height' && patient) {
    const zResult = heightForAgeZScore(observation.value, patient.date_of_birth, patient.gender, new Date(observation.effective_date));
    if (zResult.zscore < config.ZSCORE_EXTREME_LOW || zResult.zscore > config.ZSCORE_EXTREME_HIGH) {
      alerts.push({
        alert_type: 'R008_ZSCORE_EXTREME',
        severity: config.SEVERITY.HIGH,
        message: `Height z-score = ${zResult.zscore} (${zResult.interpretation}). Extreme outlier for age`,
        explanation: `A z-score of ${zResult.zscore} means this height is extremely far from the WHO reference median for this age and gender.`
      });
    }
  }

  // R006: Weight velocity
  if (observation.type === 'weight' && history && history.length > 0) {
    const weightHistory = history
      .filter(h => h.type === 'weight')
      .sort((a, b) => new Date(b.effective_date) - new Date(a.effective_date));

    if (weightHistory.length > 0) {
      const prev = weightHistory[0];
      const daysDiff = (new Date(observation.effective_date) - new Date(prev.effective_date)) / (1000 * 60 * 60 * 24);

      if (daysDiff > 0) {
        const percentChange = Math.abs(observation.value - prev.value) / prev.value * 100;
        const percentPerDay = percentChange / daysDiff;

        if (percentPerDay > config.MAX_WEIGHT_CHANGE_PERCENT_PER_DAY) {
          const direction = observation.value > prev.value ? 'gain' : 'loss';
          alerts.push({
            alert_type: 'R006_IMPLAUSIBLE_VELOCITY',
            severity: config.SEVERITY.HIGH,
            message: `Implausible weight ${direction}: ${prev.value}→${observation.value} kg (${percentChange.toFixed(1)}% in ${daysDiff.toFixed(0)} days = ${percentPerDay.toFixed(1)}%/day)`,
            explanation: `Weight changed by ${percentChange.toFixed(1)}% over ${daysDiff.toFixed(0)} days (${percentPerDay.toFixed(1)}% per day). Expected maximum is ${config.MAX_WEIGHT_CHANGE_PERCENT_PER_DAY}% per day. This rapid ${direction} may indicate a measurement error.`
          });
        }
      }
    }
  }

  // R009: Height decrease
  if (observation.type === 'height' && history && history.length > 0) {
    const heightHistory = history
      .filter(h => h.type === 'height')
      .sort((a, b) => new Date(b.effective_date) - new Date(a.effective_date));

    if (heightHistory.length > 0) {
      const prev = heightHistory[0];
      const ageMonths = ageInMonths(patient.date_of_birth, new Date(observation.effective_date));

      // Height should not decrease significantly for children (small measurement variance is OK)
      if (observation.value < prev.value - 1.0 && ageMonths < 216) {
        alerts.push({
          alert_type: 'R009_HEIGHT_DECREASE',
          severity: config.SEVERITY.HIGH,
          message: `Height decreased: ${prev.value}→${observation.value} cm (−${(prev.value - observation.value).toFixed(1)} cm). Height should not decrease in growing children`,
          explanation: `A decrease in height from ${prev.value} cm to ${observation.value} cm is physiologically impossible for a growing child. This likely indicates a measurement error.`
        });
      }
    }
  }

  return alerts;
}

module.exports = { checkTrajectory };
